### PR TITLE
Additional tests for BIP 174

### PR DESCRIPTION
**Pull request by achow101**

### CODE REVIEW

**Comment by jb55** [comment]
I was meaning to mention this on the previous PRs. Is there any point to including the base64 encoding as well? The translation from hex is pretty straightforward. Just thinking it would cut down on the verbosity a bit.

**Comment by achow101** [comment]
> Is there any point to including the base64 encoding as well? The translation from hex is pretty straightforward.

The point was to effectively provide the two forms that a PSBT would be found in: as a binary file or as a Base64 string. It helps those who are writing tests to know exactly what to expect especially since there are (apparently) multiple ways of doing Base 64.

**Comment by jb55** [comment]
> there are (apparently) multiple ways of doing Base 64

https://en.wikipedia.org/wiki/Base64#Variants_summary_table

so it seems... seems reasonable to include it in that case...

**Comment by achow101** [comment]
@luke-jr This is ready.

**Comment by achow101** [comment]
I updated this with some parts that specify what signers should look out for and an algorithm for a simple signer. Also more tests for those.

File: bip-0174.mediawiki [hunk 1]
```diff
@@ -305,6 +305,47 @@ Signers do not need to sign for all possible input types. For example, a signer
 
 A single entity is likely to be both a Signer and an Updater as it can update a PSBT with necessary information prior to signing it.
 
+====Data Signers Check For====
+
+For a Signer to only produce valid signatures for what it expects to sign, it must check that the following conditions are true:
+
+* If a non-witness UTXO is provided, it's hash must match the hash specified in the prevout
```
Comment (sipa) [inline bip-0174.mediawiki]: typo: "its".

File: bip-0174.mediawiki [hunk 2]
```diff
@@ -305,6 +305,47 @@ Signers do not need to sign for all possible input types. For example, a signer
...
+            sign(witness_sighash(script_code, i, input))
+
+sign_non_witness(script_code, i):
+    for key in psbt.inputs[i].keys:
+        if IsMine(key):
+            sign(non_witness_sighash(script_code, i, input))
+
+for input,i in enumerate(psbt.inputs):
+    if non_witness_utxo.exists:
+        assert(sha256d(non_witness_utxo) == prevhash)
```
Comment (sipa) [inline bip-0174.mediawiki]: write `prevhash` as `psbt.tx.input[i].prevout.hash`?

File: bip-0174.mediawiki [hunk 3]
```diff
@@ -305,6 +305,47 @@ Signers do not need to sign for all possible input types. For example, a signer
...
+        if redeemScript.exists:
+            assert(non_witness_utxo.scriptPubKey == P2SH(redeemScript))
```
Comment (sipa) [inline bip-0174.mediawiki]: Write as `non_witness_utxo.vout[psbt.tx.input[i].prevout.n].scriptPubKey` ?

File: bip-0174.mediawiki [hunk 4]
```diff
@@ -305,6 +305,53 @@ Signers do not need to sign for all possible input types. For example, a signer
 
 A single entity is likely to be both a Signer and an Updater as it can update a PSBT with necessary information prior to signing it.
 
+====Data Signers Check For====
+
+For a Signer to only produce valid signatures for what it expects to sign, it must check that the following conditions are true:
+
+* If a non-witness UTXO is provided, it's hash must match the hash specified in the prevout
+* If a witness UTXO is provided, a witness signature is being produced
```
Comment (sipa) [inline bip-0174.mediawiki]: Perhaps write this as "If a witness UTXO is provided, no non-witness signature may be created"? The current language sounds like a signature must be created.

File: bip-0174.mediawiki [hunk 5]
```diff
@@ -305,6 +305,53 @@ Signers do not need to sign for all possible input types. For example, a signer
...
+    if non_witness_utxo.exists:
+        assert(sha256d(non_witness_utxo) == prevhash)
+        if redeemScript.exists:
+            assert(non_witness_utxo.scriptPubKey == P2SH(redeemScript))
+            sign_non_witness(redeemScript)
+        else:
+            sign_non_witness(non_witness_utxo.scriptPubKey)
+    else if witness_utxo.exists:
+        if redeemScript.exists:
+            assert(witness_utxos.scriptPubKey == P2SH(redeemScript))
```
Comment (sipa) [inline bip-0174.mediawiki]: Typo: "witness_utxos".

File: bip-0174.mediawiki [hunk 6]
```diff
@@ -305,6 +305,53 @@ Signers do not need to sign for all possible input types. For example, a signer
...
+            script = redeemScript
+        else:
+            script = witness_utxo.scriptPubKey
+        if IsP2WPKH(script):
+            sign_witness(P2PKH(script))
```
Comment (sipa) [inline bip-0174.mediawiki]: You can't really convert a script to P2PKH. What about `P2PKH(script[2:])`?

File: bip-0174.mediawiki (hunk 1 above)
Comment (achow101) [reply bip-0174.mediawiki]: Done

File: bip-0174.mediawiki (hunk 2 above)
Comment (achow101) [reply bip-0174.mediawiki]: Done

File: bip-0174.mediawiki (hunk 3 above)
Comment (achow101) [reply bip-0174.mediawiki]: Done

File: bip-0174.mediawiki (hunk 4 above)
Comment (achow101) [reply bip-0174.mediawiki]: Done

File: bip-0174.mediawiki (hunk 5 above)
Comment (achow101) [reply bip-0174.mediawiki]: Fixed

File: bip-0174.mediawiki (hunk 6 above)
Comment (achow101) [reply bip-0174.mediawiki]: Done

**Pull request merged**